File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/IntroCourseLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.core.navigation.MobilePlatformFeatures

/** Intro course onboarding helpers (IC07). */
object IntroCourseLogic {
    fun introCourseEnabled(features: MobilePlatformFeatures): Boolean = features.introCourseEnabled

    fun cacheKeyProgress(): String = "intro-course:progress"

    fun cardState(progress: IntroCourseProgress?, loading: Boolean, error: Boolean): IntroCourseCardState {
        if (loading) return IntroCourseCardState.Loading
        if (error || progress == null) return IntroCourseCardState.Error
        if (!progress.enrolled) return IntroCourseCardState.Hidden
        if (progress.completedAt != null) return IntroCourseCardState.Completed
        if (progress.modulesComplete <= 0) return IntroCourseCardState.NotStarted
        return IntroCourseCardState.InProgress
    }

    fun shouldShowCelebration(progress: IntroCourseProgress?): Boolean {
        if (progress?.enrolled != true || progress.completedAt == null) return false
        return progress.celebrationSeen != true
    }

    fun fallbackRoute(courseCode: String = IntroCourseConstants.courseCode): String =
        "/courses/$courseCode"

    fun ctaRoute(progress: IntroCourseProgress): String =
        progress.nextItem?.route ?: fallbackRoute(progress.courseCode ?: IntroCourseConstants.courseCode)

    fun isIntroCourse(courseCode: String): Boolean = courseCode == IntroCourseConstants.courseCode
}

File: clients/ios/Lextures/Core/LMS/IntroCourseLogic.swift
import Foundation

/// Intro course onboarding helpers (IC07; mirrors web `intro-course-api.ts`).
enum IntroCourseLogic {
    static func introCourseEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.introCourseEnabled
    }

    static func cacheKeyProgress() -> String { "intro-course:progress" }

    static func cardState(
        progress: IntroCourseProgress?,
        loading: Bool,
        error: Bool
    ) -> IntroCourseCardState {
        if loading { return .loading }
        if error || progress == nil { return .error }
        if !progress!.enrolled { return .hidden }
        if progress!.completedAt != nil { return .completed }
        if progress!.modulesComplete <= 0 { return .notStarted }
        return .inProgress
    }

    static func shouldShowCelebration(_ progress: IntroCourseProgress?) -> Bool {
        guard let progress, progress.enrolled, progress.completedAt != nil else { return false }
        return progress.celebrationSeen != true
    }

    static func fallbackRoute(courseCode: String = IntroCourseConstants.courseCode) -> String {
        "/courses/\(courseCode)"
    }

    static func ctaRoute(for progress: IntroCourseProgress) -> String {
        progress.nextItem?.route ?? fallbackRoute(courseCode: progress.courseCode ?? IntroCourseConstants.courseCode)
    }

    static func isIntroCourse(_ courseCode: String) -> Bool {
        courseCode == IntroCourseConstants.courseCode
    }
}